### PR TITLE
Adding gateway counters to Agrona

### PR DESCRIPTION
**Pull request by irina-mitrea-luxoft**
Needs Java 1.8 in order to build.

### CODE REVIEW

**Comment by brennangaunce** [comment]
I'm seeing the following when I build and run:

```
kaazing-gateway-community-develop-SNAPSHOT> export GATEWAY_OPTS=-Dorg.kaazing.gateway.management.AGRONA_ENABLED=true ; bin/gateway.start
INFO  Kaazing Gateway (develop-SNAPSHOT)
INFO  Configuration file: /Users/bren/dve/gateway-agrona/alin/gateway/distribution/target/tar/kaazing-gateway-community-develop-SNAPSHOT/conf/gateway-config-minimal.xml
WARN  Error while injecting named configuration resource
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.kaazing.gateway.server.context.resolve.GatewayContextResolver.injectResources(GatewayContextResolver.java:1273)
	at org.kaazing.gateway.server.context.resolve.GatewayContextResolver.injectResources(GatewayContextResolver.java:1252)
	at org.kaazing.gateway.server.context.resolve.GatewayContextResolver.resolve(GatewayContextResolver.java:277)
	at org.kaazing.gateway.server.impl.GatewayImpl.launch(GatewayImpl.java:303)
	at org.kaazing.gateway.server.GatewayCommandLineProcessor.launchGateway(GatewayCommandLineProcessor.java:105)
	at org.kaazing.gateway.server.GatewayCommandLineProcessor.launchGateway(GatewayCommandLineProcessor.java:51)
	at org.kaazing.gateway.server.Main.main(Main.java:43)
Caused by: java.lang.IllegalArgumentException: could not create Monitoring directory directory: /kaazing
	at uk.co.real_logic.agrona.IoUtil.ensureDirectoryExists(IoUtil.java:127)
	at org.kaazing.gateway.management.monitoring.configuration.impl.AgronaMonitoringEntityFactoryBuilder.getMonitoringDirName(AgronaMonitoringEntityFactoryBuilder.java:102)
	at org.kaazing.gateway.management.monitoring.configuration.impl.AgronaMonitoringEntityFactoryBuilder.build(AgronaMonitoringEntityFactoryBuilder.java:52)
	at org.kaazing.gateway.management.monitoring.configuration.impl.MonitoringEntityFactoryInjectorImpl.makeMonitoringEntityFactory(MonitoringEntityFactoryInjectorImpl.java:49)
	at org.kaazing.gateway.management.context.DefaultManagementContext.buildMonitoringEntityFactory(DefaultManagementContext.java:658)
	at org.kaazing.gateway.management.context.DefaultManagementContext.setConfiguration(DefaultManagementContext.java:298)
	... 11 more
INFO  Stopping server
INFO  Stopping services
INFO    http://brennans-macbook-pro-2.local:8000/
INFO    http://brennans-macbook-pro-2.local:8000/commandcenter
INFO    ws://brennans-macbook-pro-2.local:8000/echo
INFO    ws://brennans-macbook-pro-2.local:8000/snmp
INFO  Stopped services
INFO  Stopped server successfully in 0.000 secs at 2015-05-15 14:03:19
ERROR Error starting Gateway: caught exception java.lang.NullPointerException
ERROR Error starting Gateway: caught exception java.lang.NullPointerException
ERROR Gateway failed to launch
java.lang.NullPointerException
	at org.kaazing.gateway.management.filter.ManagementFilter.<init>(ManagementFilter.java:58)
	at org.kaazing.gateway.management.context.DefaultManagementContext.addManagementFilter(DefaultManagementContext.java:641)
	at org.kaazing.gateway.management.context.DefaultManagementContext.getManagementFilter(DefaultManagementContext.java:669)
	at org.kaazing.gateway.management.context.ManagementGatewayListener.addSessionInitializer(ManagementGatewayListener.java:80)
	at org.kaazing.gateway.management.context.ManagementGatewayListener.startingService(ManagementGatewayListener.java:67)
	at org.kaazing.gateway.server.GatewayObserver.startingService(GatewayObserver.java:104)
	at org.kaazing.gateway.server.Launcher.initInternal(Launcher.java:83)
	at org.kaazing.gateway.server.Launcher.init(Launcher.java:54)
	at org.kaazing.gateway.server.impl.GatewayImpl.launch(GatewayImpl.java:308)
	at org.kaazing.gateway.server.GatewayCommandLineProcessor.launchGateway(GatewayCommandLineProcessor.java:105)
	at org.kaazing.gateway.server.GatewayCommandLineProcessor.launchGateway(GatewayCommandLineProcessor.java:51)
	at org.kaazing.gateway.server.Main.main(Main.java:43)
java.lang.NullPointerException
	at org.kaazing.gateway.management.filter.ManagementFilter.<init>(ManagementFilter.java:58)
	at org.kaazing.gateway.management.context.DefaultManagementContext.addManagementFilter(DefaultManagementContext.java:641)
	at org.kaazing.gateway.management.context.DefaultManagementContext.getManagementFilter(DefaultManagementContext.java:669)
	at org.kaazing.gateway.management.context.ManagementGatewayListener.addSessionInitializer(ManagementGatewayListener.java:80)
	at org.kaazing.gateway.management.context.ManagementGatewayListener.startingService(ManagementGatewayListener.java:67)
	at org.kaazing.gateway.server.GatewayObserver.startingService(GatewayObserver.java:104)
	at org.kaazing.gateway.server.Launcher.initInternal(Launcher.java:83)
	at org.kaazing.gateway.server.Launcher.init(Launcher.java:54)
	at org.kaazing.gateway.server.impl.GatewayImpl.launch(GatewayImpl.java:308)
	at org.kaazing.gateway.server.GatewayCommandLineProcessor.launchGateway(GatewayCommandLineProcessor.java:105)
	at org.kaazing.gateway.server.GatewayCommandLineProcessor.launchGateway(GatewayCommandLineProcessor.java:51)
	at org.kaazing.gateway.server.Main.main(Main.java:43)
INFO  Stopping server
INFO  Stopping services
INFO    http://brennans-macbook-pro-2.local:8000/
INFO    http://brennans-macbook-pro-2.local:8000/commandcenter
INFO    ws://brennans-macbook-pro-2.local:8000/echo
INFO    ws://brennans-macbook-pro-2.local:8000/snmp
INFO  Stopped services
INFO  Stopped server successfully in 0.000 secs at 2015-05-15 14:03:19
```

File: distribution/src/main/assembly/generic-bin.xml [hunk 1]
```diff
@@ -96,6 +96,7 @@
       <outputDirectory>lib</outputDirectory>
       <useProjectArtifact>false</useProjectArtifact>
       <includes>
+	    <include>uk.co.real-logic:Agrona</include>
```
Comment (krismcqueen) [inline distribution/src/main/assembly/generic-bin.xml]: no tabs please

File: management/src/main/java/org/kaazing/gateway/management/context/DefaultManagementContext.java [hunk 2]
```diff
@@ -620,11 +638,28 @@ private ServiceDefaultsConfigurationBean addServiceDefaultsConfigurationBean(Ser
     }
 
     private ManagementFilter addManagementFilter(ServiceContext serviceContext, ServiceManagementBean serviceBean) {
-        ManagementFilter managementFilter = new ManagementFilter(serviceBean);
+        ManagementFilter managementFilter = new ManagementFilter(serviceBean,
+                                                                 monitoringEntityFactory,
```
Comment (krismcqueen) [inline management/src/main/java/org/kaazing/gateway/management/context/DefaultManagementContext.java]: What happens if monitoringEntityFactory is null here?  Maybe call a different constructor?

File: management/src/main/java/org/kaazing/gateway/management/monitoring/entity/impl/AgronaMonitoringEntityFactory.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            MappedByteBuffer mappedMonitorFile,
+            File monitoringDirectory) {
+        this.countersManager = countersManager;
+        this.mappedMonitorDirectory = mappedMonitorFile;
+        this.monitoringDirectory = monitoringDirectory;
+    }
+
+    @Override
+    public BooleanMonitoringEntity makeBooleanMonitoringEntity() {
+        // TODO Implement the boolean monitoring entity when needed
```
Comment (krismcqueen) [inline management/src/main/java/org/kaazing/gateway/management/monitoring/entity/impl/AgronaMonitoringEntityFactory.java]: If it's not needed now (and the feature is not in the next sprint or two) then I suggest deleting this method.

File: management/src/main/java/org/kaazing/gateway/management/monitoring/entity/impl/AgronaMonitoringEntityFactory.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        counters.add(counter);
+
+        LongMonitoringCounter longMonitoringCounter = new AgronaLongMonitoringCounter(counter);
+
+        return longMonitoringCounter;
+    }
+
+    @Override
+    public StringMonitoringEntity makeStringMonitoringEntity() {
+        // TODO Implement the String monitoring entity when needed
```
Comment (krismcqueen) [inline management/src/main/java/org/kaazing/gateway/management/monitoring/entity/impl/AgronaMonitoringEntityFactory.java]: If it's not needed now (and the feature is not in the next sprint or two) then I suggest deleting this method.

File: management/src/test/java/org/kaazing/gateway/management/monitoring/configuration/impl/MonitoringEntityFactoryInjectorTest.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        MonitoringEntityFactoryInjector injector = new MonitoringEntityFactoryInjectorImpl(new Properties());
+        MonitoringEntityFactory monitoringEntityFactory = injector.makeMonitoringEntityFactory();
+        assertTrue(monitoringEntityFactory instanceof DefaultMonitoringEntityFactoryStub);
+    }
+
+    @Test
+    public void testMonitoringEntityFactoryInjectorAgronaEnabledFalse() {
+        Properties configuration = new Properties();
+        configuration.setProperty(AGRONA_ENABLED, Boolean.toString(false));
+		MonitoringEntityFactoryInjector injector = new MonitoringEntityFactoryInjectorImpl(configuration);
```
Comment (krismcqueen) [inline management/src/test/java/org/kaazing/gateway/management/monitoring/configuration/impl/MonitoringEntityFactoryInjectorTest.java]: tabs to spaces...

File: management/src/test/java/org/kaazing/gateway/management/monitoring/configuration/impl/MonitoringEntityFactoryInjectorTest.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        MonitoringEntityFactory monitoringEntityFactory = injector.makeMonitoringEntityFactory();
+        assertTrue(monitoringEntityFactory instanceof DefaultMonitoringEntityFactoryStub);
+    }
+
+    @Test
+    public void testMonitoringEntityFactoryInjectorAgronaEnabledTrue() {
+        Properties configuration = new Properties();
+        configuration.setProperty(AGRONA_ENABLED, Boolean.toString(true));
+		MonitoringEntityFactoryInjector injector = new MonitoringEntityFactoryInjectorImpl(configuration);
```
Comment (krismcqueen) [inline management/src/test/java/org/kaazing/gateway/management/monitoring/configuration/impl/MonitoringEntityFactoryInjectorTest.java]: tabs to spaces

**Comment by krismcqueen** [comment]
Just some minor feedback.  Otherwise looks good.

File: management/src/main/java/org/kaazing/gateway/management/monitoring/entity/impl/AgronaMonitoringEntityFactory.java (hunk 3 above)
Comment (irina-mitrea-luxoft) [inline management/src/main/java/org/kaazing/gateway/management/monitoring/entity/impl/AgronaMonitoringEntityFactory.java]: Due to the fact that we first need to establish the priorities for the upcoming sprints, we propose that we keep this implementation for now and wait until a decision has been made regarding this.

**Comment by brennangaunce** [comment]
Is the build failing because we now require JDK 8?

**Comment by krismcqueen** [comment]
@irina-mitrea-luxoft As soon as build is passing this can be merged.  Please send me a ping when it's ready and building successfully.  Thanks!

**Comment by irina-mitrea-luxoft** [comment]
The Agrona-related unit tests currently fail due to the fact that we now require Java 8 in order to perform the build.

**Comment by irina-mitrea-luxoft** [comment]
Java 8 is now required in order to build the gateway, the build will fail with Java 7.

**Comment by irina-mitrea-luxoft** [comment]
Hello,

Java 8 is now required in order to build the gateway, the build will fail with Java 7.

**Comment by dpwspoon** [comment]
Please update the Travis.yml from

```
language: Java
jdk:
  - oraclejdk7
  - openjdk7
install: mvn clean -DskipTests -DskipITs install
script: mvn clean verify
after_success:
  - mvn clean cobertura:cobertura-integration-test coveralls:report
```

To 
```
language: Java
jdk:
  - oraclejdk8
  - openjdk8
install: mvn clean -DskipTests -DskipITs install
script: mvn clean verify
after_success:
  - mvn clean cobertura:cobertura-integration-test coveralls:report
```

The build should then pass ;-)

**Comment by irina-mitrea-luxoft** [comment]
Removed openjdk8 from .travis.yml. Reference: https://github.com/travis-ci/travis-ci/issues/1647

**Comment by dpwspoon** [comment]
LGTM

**Comment by dpwspoon** [comment]
Check that, as part of updating to Java 8 we also need to do the following:

1. update the README with the minimum java version
2. ignore javadoc lint (needed for the release, test with `mvn javadoc:javadoc`).  See [this](http://blog.joda.org/2014/02/turning-off-doclint-in-jdk-8-javadoc.html).  And file an issue to fix the javadoc linting issues in the future

**Comment by brennangaunce** [comment]
David - Irina is now on vacation.  Please merge this once you have confirmed Java 8 issues have been addressed (looks like the build completed successfully)

**Comment by brennangaunce** [comment]
See also https://github.com/kaazing/gateway/pull/158

**Pull request merged**